Library for managing a process

gsagula opened the pull request:
In this PR:

- Created a struct that wraps exec/cmd with a few extras functions for persisting/reading stdout/stderr.
- Allows seeing the output of a running process while saving it to a file.
- Creates a unique id for tracking the process.
- Allows context cancellation for supporting service termination.

Signed-off-by: Gabriel Linden Sagula <glindens@bitfield.org>

gsagula commented on internal/processor.go:
```diff
@@ -0,0 +1,201 @@
+// stderr to a file in the relative path. It takes a context object, a
+// string command and returns P or error. The command string is assumed
+// to contain binary path and GNU-style arguments, e.g. "/usr/bin/ls -la"
+// TODO: This function should also take env vars.
```
This TODO can be removed.

NajiObeid commented on internal/processor.go:
```diff
@@ -0,0 +1,201 @@
+)
+
+// P is an object that manages a child process.
+type P struct {
```
I'd usually prefer more a descriptive name.
That also applies to all the single letter variables down below, with the exception of functions with receiver types, loop indices or wherever it would never cause any confusion

NajiObeid commented on internal/processor.go:
```diff
@@ -0,0 +1,201 @@
+
+// P is an object that manages a child process.
+type P struct {
+	id     string
```
I'm assuming this would eventually be exported?

gsagula commented on internal/processor.go:
```diff
@@ -0,0 +1,201 @@
+	return p.id
+}
+
+func (p *P) wait(ctx context.Context, w io.WriteCloser) {
```
We should defer a context cancellation here.

NajiObeid commented on internal/processor.go:
```diff
@@ -0,0 +1,201 @@
+// TODO: This function should also take env vars.
+func Process(ctx context.Context, command string, env ...string) (*P, error) {
+	s := strings.Split(command, " ")
+	if len(s) < 1 && len(s[0]) < 1 {
```
in this case `len(s)` will always be >= 1

NajiObeid commented on internal/processor.go:
```diff
@@ -0,0 +1,201 @@
+// string command and returns P or error. The command string is assumed
+// to contain binary path and GNU-style arguments, e.g. "/usr/bin/ls -la"
+// TODO: This function should also take env vars.
+func Process(ctx context.Context, command string, env ...string) (*P, error) {
```
I would prefer if this was called `NewProcess` or `StartProcess` or something more descriptive

NajiObeid commented on internal/processor.go:
```diff
@@ -0,0 +1,201 @@
+}
+
+func createDir(dir string, perm os.FileMode) error {
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
```
This is not needed as `MkdirAll` already does check if the directory already exists

bernardjkim commented on internal/processor.go:
```diff
@@ -0,0 +1,201 @@
+				}
+				break loop
+			}
+			w.Write(chunk[:n])
```
Is it safe to ignore any potential errors returned here?

gsagula replied:
Ha, no! I forgot about this one. I will handle the error and the context cancellation for this case.

bernardjkim commented on internal/processor.go:
```diff
@@ -0,0 +1,201 @@
+	if p.cmd.ProcessState != nil {
+		return p.cmd.ProcessState.ExitCode(), nil
+	}
+	return -100, errors.New("process has not exited or it has been terminated")
```
nit: Any particular reason to return -100 exit code here?

gsagula replied:
I know it looks weird but I didn't want to return 0 or any other code that the ExitCode() may return. I will add a TODO here to revisit this design or if you have a suggestion, I can change it now.

gsagula replied:
I change it so the it returns a reference to an int. In case of error, it will return nil.

bernardjkim reviewed (approved):
Doesn't look like the status of a job is being tracked. If you plan on implementing this in a separate PR that's fine.

gsagula commented:
@bernardjkim @NajiObeid Thank you for the review. I just noticed that there is a small problem with the cancellation. I fix that and also address your comments.

NajiObeid commented on internal/processor.go:
```diff
@@ -0,0 +1,201 @@
+	}
+
+	p.reader = io.MultiReader(stdout, stderr)
+	if p.reader == nil {
```
just curious will this ever be nil?

gsagula replied:
No, it will be never nil. I just wanted to be more explicit about it in case myself or someone else decides to change this code.

NajiObeid commented on internal/processor.go:
```diff
@@ -0,0 +1,201 @@
+		return nil, err
+	}
+
+	err = createDir(p.dirOut(), 0755)
```
what would happen to the command if we fail here?

gsagula replied:
It would leak a process. I've already changed it when fixing the other cancellation issue that I mentioned earlier.

knisbet commented on internal/processor.go:
```diff
@@ -0,0 +1,201 @@
+// to contain binary path and GNU-style arguments, e.g. "/usr/bin/ls -la"
+// TODO: This function should also take env vars.
+func Process(ctx context.Context, command string, env ...string) (*P, error) {
+	s := strings.Split(command, " ")
```
So splitting on spaces creates a bit of a gap, since normally args can accept spaces. Something like `echo "hello world" "how are you today?"` is two args to the program, one for `hello world` and a second for `how are you today?`. While not important for my contrived example, it would cause programs to get confused. For example if I were to run `bash -c "echo hello world", bash wouldn't be able to process this.

gsagula replied:
You are absolutely right. I will take an array of arguments. I will need to change the client and the protobuf to do the same.

NajiObeid commented on internal/processor.go:
```diff
@@ -0,0 +1,201 @@
+loop:
+	for {
+		select {
+		case <-ctx.Done():
```
I think this is redundant. 
If the context becomes done before the commands completes the command will be killed. The reader will then return an io.EOF or some other error and exit the loop.

gsagula replied:
That's right. Also in this loop, we should be cancelling the process if anything goes wrong. I've change it by passing a `context.CancelFunc` to wait() so we can deffer and invoke it where it matters.

NajiObeid commented on internal/processor.go:
```diff
@@ -0,0 +1,201 @@
+		log.Println("writer cannot be nil")
+		return
+	}
+	defer ErrorLogFN(w.Close)
```
where is this defined?

gsagula replied:
I messed up the squash commit before open the PR. It is in another file:
```
// ErrorLogFN is a wrapper for deffering functions that return error.
  func ErrorLogFN(f func() error, msg ...string) {
          if err := f(); err != nil {
                  log.Println(msg, err)
          }
  }
```

gsagula replied:
The missing file:

https://github.com/gsagula/remote-worker/pull/4/files#diff-630d21b610ac56c4c88441e1ea512dba613ef29c34896aa9db7ddee6f6d1899cR6

NajiObeid commented on internal/processor.go:
```diff
@@ -0,0 +1,201 @@
+	return path.Join("processes", p.id, "cmd")
+}
+
+func (p *P) dirOut() string {
+	return path.Join("processes", p.id, "out")
+}
+
+func (p *P) pathFile() string {
+	return path.Join("processes", p.id, "out", "combined.out")
+}
```
This can lead to potential problems if someone changed one of the functions and forgot the other one. It would be better if `pathFile` is based on the output of `dirOut`

gsagula commented:
@bernardjkim 

> Doesn't look like the status of a job is being tracked. If you plan on implementing this in a separate PR that's fine.

I've added this function for fetching the process status. I couldn't find anything in the standard library.

https://github.com/gsagula/remote-worker/pull/4/files#diff-ebf403516411d002fe2275d67de0863bd710d19158729cafdeae99a5e0cf56eaR125

gsagula commented:
Thank you everyone for grinding it. You will see that I moved some things around and also fixed some issues related to cancellation. I also added a missing file for a commonly used function as well as a function for fetch the state of a process. I'm not 100% sure if that's the correct way of doing it, but it seems to work fine.

NajiObeid commented on internal/processor.go:
```diff
@@ -0,0 +1,256 @@
+		return nil, err
+	}
+
+	err = p.cmd.Start()
```
in theory it would be better to keep the start and wait command in the same place/function. that way you'd have more direct control of what can happen in between and not have someone add code that inadvertently break things. 
so maybe this function can be called `prepare` and the `wait` function can be renamed to `run` or something else.
that's mostly a nitpicking, so I am ok with the current implementation.

gsagula replied:
I thought about a `prepare` function. Great suggestion! I will come back here to do this change.

NajiObeid commented on internal/processor.go:
```diff
@@ -0,0 +1,256 @@
+// Status fetches the current process state.
+func (p *Processor) Status() (string, error) {
+	if p.cmd.ProcessState != nil {
+		return "Exited", nil
```
why not add this state to the consts above?

gsagula replied:
Ha, good call. I added `exited` to this function in the last minute. Initially I want this function to be called `State` and to deal with running states only. This is another corner that I cut but I may change it later for correctness if I have time.

NajiObeid commented on internal/processor.go:
```diff
@@ -0,0 +1,256 @@
+		return UnixProcStateDead, nil
+	}
+
+	return "", fmt.Errorf("unknown unix state %c", s)
```
nit: maybe also add an unknown/undefined state const?

NajiObeid commented on internal/processor.go:
```diff
@@ -0,0 +1,256 @@
+		return "", fmt.Errorf("failed to parse /proc stat: %v", err)
+	}
+
+	switch s {
+	case 'R':
+		return UnixProcStateRunning, nil
+	case 'S':
+		return UnixProcStateInterruptible, nil
+	case 'D':
+		return UnixProcStateUninterruptible, nil
+	case 'T':
+		return UnixProcStateStopped, nil
+	case 'Z':
+		return UnixProcStateZombie, nil
+	case 'X':
+		return UnixProcStateDead, nil
+	}
```
ideally it would be better to separate that kind of logic from the code into it's own type, but it's ok for this exercise, just make sure to mention the corners you're cutting

gsagula replied:
Right, I should be more explicit. I'm just trying to wrap up the project now, but I will come back here too if I get some spear time.
